[SM-69] feature: create org-switcher, bit-nav-item, bit-nav-group, bit-nav-divider

willmartian opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR adds an `org-switcher` component for the secrets manager. The organization switcher is a higher order component composed of three new components from the generic component library,`nav-group`, `nav-item`, and `nav-divider`

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

- **`libs/components/src/navigation/nav-item.component`**
  - Generic component to display links and buttons in the vertical navigation. 
- **`libs/components/src/navigation/nav-group.component`**
  - Generic component to group multiple `nav-item`
- **`libs/components/src/navigation/nav-divider.component`**
  - Generic component to divide navigation items
- **`bitwarden_license/bit-web/src/app/secrets-manager/layout/org-switcher.component`**
  - `nav-group` of user's organizations
- **`bitwarden_license/bit-web/src/app/secrets-manager/layout/navigation.component`**
  - Updated to use `org-switcher` and `nav-item`
- **`bitwarden_license/bit-web/src/app/secrets-manager/overview/*`**
  - Skeleton page for organization overview (used with with `org-switcher`)

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

- `nav-item` in light and dark theme
<img width="270" alt="image" src="https://user-images.githubusercontent.com/17113462/203662069-ce7974e8-c0c1-4d6e-98a8-ba9ec8c3d62c.png">

- `nav-group` and `nav-item` with `variant="tree"`
<img width="409" alt="image" src="https://user-images.githubusercontent.com/17113462/204551630-7854ecfa-28e1-4949-b179-f58ae0f5b2fa.png">


- `org-switcher` and other items in vertical navigation
<img width="409" alt="image" src="https://user-images.githubusercontent.com/17113462/204545263-55746939-655e-4ddd-998f-264297e3d4e9.png">

- `org-switcher` click navigation
<video width="409" alt="image" src="https://user-images.githubusercontent.com/17113462/204667868-6753a3a2-56cd-4034-95ad-c30932cfba2e.mov"></video>

- `org-switcher` keyboard navigation and focus order

<video width="409" alt="image" src="https://user-images.githubusercontent.com/17113462/204668387-b943e7bd-384e-4d2d-966f-83483e7c161a.mov"></video>

## Screen Reader

Firefox on Mac with Voice Over

- `org-switcher` and left navigation

<video src="https://user-images.githubusercontent.com/17113462/205356166-9d52ac01-3c17-4a73-ad0e-4ae7ef503663.mov"></video>

- `nav-group` with tree

<video width="409" alt="image" src="https://user-images.githubusercontent.com/17113462/205084046-af8ce73f-6ab1-486b-a566-d6ffe7a270f5.mov"></video>

- `nav-item` with multiple sub buttons

<video width="409" alt="image" src="https://user-images.githubusercontent.com/17113462/205084088-96918827-812f-4b91-8141-85da96269c6e.mov"></video>





## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

Hinton commented:
We should look into building the generic components for this functionality that follows the design specification in https://www.figma.com/file/Zt3YSeb6E6lebAffrNLa0h/Tailwind-Component-Library?node-id=4687%3A86642&t=z9DMaF7JpmhaJqjr-4. Same as the `bitNavigation` directive. (The reason it's a directive and not a component was to support both buttons and links)

In this case I believe the organization switcher falls under `Page Groups - parent page content` and `Subpages`. We also need a way to show the current active organization which most likely would fall under `Page Trees as filters` and `Sub Page Options Icons`.

willmartian commented:
> We should look into building the generic components for this functionality that follows the design specification in https://www.figma.com/file/Zt3YSeb6E6lebAffrNLa0h/Tailwind-Component-Library?node-id=4687%3A86642&t=z9DMaF7JpmhaJqjr-4. Same as the `bitNavigation` directive, which already implements the `Basic UI States`. (The reason it's a directive and not a component was to support both buttons and links)
> 
> In this case I believe the organization switcher falls under `Page Groups - parent page content` and `Subpages`. We also need a way to show the current active organization which most likely would fall under `Page Trees as filters` and `Sub Page Options Icons`.

Hey @Hinton, still working on this, but I think I am most of the way to hitting that design spec you linked with `nav-item` and `nav-group`. My goal for `nav-item` was to expand `bitNavigation` into a component is to allow it to support the nested interactive content in the design spec (drop down buttons, etc). 

Should PRs that have component library features target `master`, or is targeting `sm/secrets-feature` acceptable?

coroiu commented on libs/components/src/navigation/nav-group.component.ts:
```diff
@@ -0,0 +1,96 @@
+   */
+  @Input() icon: string;
+
+  /**
+   * Route to be passed to internal `routerLink`
+   **/
+  @Input() to: string;
```
suggestion: Name `route` instead to be consistent with existing tabs component 

https://github.com/bitwarden/clients/blob/be56d29ba0831f935f05f630397c94b27694c7fe/libs/components/src/tabs/tab-nav-bar/tab-link.component.ts#L27

and upcoming (my own PR https://github.com/bitwarden/clients/pull/3762)

https://github.com/bitwarden/clients/blob/090ff1de1ba20bf0083f902efe492e03c24475a9/apps/web/src/app/shared/components/breadcrumbs/breadcrumb.component.ts#L11-L12

Or maybe switch everything over to `to` if that's more common in the angular community
(I saw this in your screenshots regarding jsdoc and thought I might as well comment on it 😄)

willmartian replied:
Thanks @coroiu! I will switch this to `route` for consistency.

Hinton commented on libs/components/src/navigation/nav-group.component.ts:
```diff
@@ -0,0 +1,96 @@
+      return;
+    }
+
+    const nestedLeftPadding = 1.5; // rem
+    this.nestDepth$.pipe(takeUntil(this.destroy$)).subscribe((v) => {
+      this.nestedNavGroups.forEach((navGroup) => {
+        navGroup.nestDepth$.next(v + 1);
+      });
+      this.contentContainer.nativeElement.style.setProperty(
+        "--nested-left-padding",
+        `${v * nestedLeftPadding}rem`
+      );
+    });
```
I feel like we're overcomplicating things. Having to access `nativeElement` is usually a bad sign.

Could we not just have a `@input` which denotes which nesting it has, and propagate that down? Which would set a simple padding for each element.

willmartian replied:
Done. Not quite sure why I didn't just bind to an `@input` to begin with. Also, I forgot one can bind to a CSS variable with `[style.--my-cool-var]=""`, so that has also been changed. 

Thanks!

Hinton commented on libs/components/src/navigation/nav-item.component.ts:
```diff
@@ -0,0 +1,81 @@
+   * - is `true` if the host component has a descendant that matches `.fvw:focus-visible`
+   * - can be removed when `:has()` or `:focus-visible-within` get broader browser support
+   */
+  protected focusVisibleWithin$ = new BehaviorSubject(false);
+  protected fvwStyles$ = this.focusVisibleWithin$.pipe(
+    map((value) => (value ? "tw-z-10 tw-rounded tw-outline-none tw-ring tw-ring-text-alt2" : ""))
+  );
+  @HostListener("focusin", ["$event.target"])
+  onFocusIn(target: HTMLElement) {
+    this.focusVisibleWithin$.next(target.matches(".fvw:focus-visible"));
+  }
+  @HostListener("focusout")
+  onFocusOut() {
+    this.focusVisibleWithin$.next(false);
+  }
```
What are we trying to do here?

coroiu commented:
As part of the vertical navigation refresh we're interested in this component as well, so I'm keeping my on this :)

Not sure if this is gonna help with some inspiration (feel free to ignore) but I actually implemented a very rudimentary version of this side menu as part of my BEEEP a few weeks ago. 
PR: https://github.com/bitwarden/clients/pull/4007
Storybook: https://62a88a6de5b807fa98886113-swgasvqwuk.chromatic.com/?path=/story/component-library-side-menu--personal-vault

It uses a slightly different approach with a complex object as input which I was never really a fan of. It also has a lot worse naming compared to what you've done here in your PR.

Edit:
I've taken a look at your PR again and it's a lot more mature than whatever I wrote so don't put too much weight on it

jtouchstonebw commented:
@willmartian thanks for including those screen recordings! Is there a way we can see what it is like with Voice Over or another screen reader? Either with audio or with text captions. Thanks!

Hinton commented on libs/components/src/navigation/nav-group.component.html:
```diff
@@ -0,0 +1,46 @@
+    ></button>
+  </ng-template>
+
+  <!-- If `tree` variant, use `slot-slart` -->
```
I try to avoid comments that describe the same thing as the code, in this case this comment says exactly the same thing as the next line of code.

Maybe remove the 2nd comment and just have something like this.

```suggestion
  <!-- Show toggle to the left for trees otherwise to the right -->
```

Hinton commented on libs/components/src/navigation/nav-group.component.html:
```diff
@@ -0,0 +1,46 @@
+  </ng-container>
+</bit-nav-item>
+
+<!-- [attr.id] needs to be a unique ID for this element -->
```
Why? Is it to connect it with aria-controls above? If so I think that would make a better comment than this.

Hinton commented on libs/components/src/navigation/nav-item.component.html:
```diff
@@ -0,0 +1,78 @@
+<div
+  class="tw-relative {{
+    active && !hideActiveStyles ? 'tw-bg-background-alt4' : 'tw-bg-background-alt3'
+  }} {{ fvwStyles$ | async }}"
```
We can `ngClass` which is a bit cleaner, or we could look into moving this to the typescript code.

```suggestion
  class="tw-relative"
  [ngClass]="[
    active && !hideActiveStyles ? 'tw-bg-background-alt4' : 'tw-bg-background-alt3',
    fvwStyles$ | async
  ]"
```

Hinton commented on libs/components/src/navigation/nav-item.component.html:
```diff
@@ -0,0 +1,78 @@
+  >
+    <div
+      #slotStart
+      class="[&>*:focus-visible::before]:!tw-ring-[white] [&>*]:!tw-text-alt2 [&>*:hover]:!tw-border-text-alt2"
```
I don't suppose `tw-ring-text-alt-2` would work?

Hinton commented on libs/components/src/navigation/nav-item.component.html:
```diff
@@ -0,0 +1,78 @@
+<div
+  class="tw-relative {{
+    active && !hideActiveStyles ? 'tw-bg-background-alt4' : 'tw-bg-background-alt3'
```
It would be nice if we could extract this to a helper, we never care about any of these individually, so not exposing them to the template avoids anyone potentially forgetting to also include `hideActiveStyles`.

`active && !hideActiveStyles`

willmartian replied:
Changed `active` to a private field and moved this logic into a getter `showActiveStyles`

Hinton reviewed (commented):
Looks good overall, some minor comments and we should be good.

willmartian commented on libs/components/src/navigation/index.ts:
```diff
@@ -1 +1,4 @@
 export * from "./navigation.module";
+export * from "./nav-divider.component";
+export * from "./nav-group.component";
+export * from "./nav-item.component";
```
Good catch 👍

willmartian commented on libs/components/src/navigation/nav-item.component.html:
```diff
@@ -0,0 +1,78 @@
+      <i class="bwi bwi-fw tw-text-white tw-mx-1 {{ icon }}"></i
+      ><span [ngClass]="active && !hideActiveStyles ? 'tw-font-bold' : 'tw-font-semibold'">{{
+        text
+      }}</span>
```
This looks like it is automatically applied by the linter

willmartian commented:
I am going to create a seperate follow up ticket for adding animations to the `nav-group` button and content reveal.

jtouchstonebw commented:
> o mind: the open/close buttons for `nav-group` reads aloud as "toggle collapse". Should this also read the label of the item it is closing? For example: "toggle collapse [Menu item name]"

Yes, if possible. I was also thinking what if we put sr-only text before the org name that says "Organization". Not sure if this is possible but having that sr-only text passed to the toggle collapse as you suggest.

Another thought, how do we handle localization of sr-only text? Or do we at all?

jtouchstonebw commented:
Also @danielleflinn may have some thoughts about the third recording with screen reader interactions for the menu items with multiple icon buttons. It looks like they don't have aria labels yet. Is that because they will be configurable at the component level?

willmartian commented:
@jtouchstonebw @danielleflinn 

> Yes, if possible. I was also thinking what if we put sr-only text before the org name that says "Organization". Not sure if this is possible but having that sr-only text passed to the toggle collapse as you suggest.

I will look into that. I am not sure if we can have sr-only text for links, but we definitely can for the toggle/icon buttons.

---

> Another thought, how do we handle localization of sr-only text? Or do we at all?

We should be able to localize any sr-only text the same way that we localize visible text. We have a localization utility, `i18n`, that we use in the HTML to substitute a keyword for its localized content...
```html
<button>{{ 'myButtonCopy' |  i18n }}</button>
```
 ...and beyond visible text, we can also use it for accessibility properties, like `aria-label`:

```html
<button aria-label="{{ 'myButtonLabel' |  i18n }}">...</button>
```

---

> Also @danielleflinn may have some thoughts about the third recording with screen reader interactions for the menu items with multiple icon buttons. It looks like they don't have aria labels yet. Is that because they will be configurable at the component level?

Yes, those are configurable at the component level with `aria-label`. I should have updated that in the Storybook before recording.

danielleflinn commented:
@jtouchstonebw 
> Another thought, how do we handle localization of sr-only text? Or do we at all?

I don't believe we translate sr-only text or aria-labels. 

@willmartian A few things on the "With Slots" story:

- The current thinking is that if a "with slot" navigation item is selected, it filters the vault items displayed to the left. With this in mind, it seems this navigation item:
  - should default to using a `<button>` directive
  - if the full element is focused, a screen reader should announce whether or not it is selected
  - if it is selected, the edit button and checkmark icon would appear (checkmark icon not being a button, but a visual indicator on whether or not the item is selected).  
- can we update the prefix icon-button to `bwi-angle-right` with the same `title` and `aria-label` behavior as the "tree" variant? I don't believe we have any other icon-button besides the tree behavior planned as a prefix for the navigation items. 
- When the edit icon is focused, can we have the screen reader announce "Edit [nav item label]"?

With that said, we could likely address these improvements in a separate PR as I think they might be a bit out of scope from the intent of your initial work. Welcome your's and @Hinton's input on that.

Hinton reviewed (approved):
I think this looks good code-wise, we might want to solve designs suggestions before merging. Or have a fast followup.

willmartian commented:
Updated the component to take a new `ariaLabel` input to enable the screen-reader behavior described by @jtouchstonebw above. I also updated the `org-switcher` to use the `ariaLabel` attribute, and prefix each item with "Organization"

<video src="https://user-images.githubusercontent.com/17113462/205356166-9d52ac01-3c17-4a73-ad0e-4ae7ef503663.mov"></video>

---

Based on discussion with discussion with @danielleflinn, selection/filtering logic will be added in a future PR. This PR adds everything visually needed for selection logic, but no actual checkbox-like functionality.

Hinton reviewed (approved):
This looks good, thanks for iterating on this.

Hinton commented on bitwarden_license/bit-web/src/app/secrets-manager/layout/org-switcher.component.html:
```diff
@@ -0,0 +1,49 @@
+      [text]="org.name"
+      [ariaLabel]="['organization' | i18n, org.name].join(' ')"
+      [route]="['../', org.id]"
+      (mainContentClicked)="parentGroup.toggle()"
```
Will this work now that toggle is private, or should we look into exposing a input for it?

willmartian replied:
Thanks for catching this. I exposed an input for it. That feels better than `(mainContentClicked)="parentGroup.open = !parentGroup.open"`
